### PR TITLE
Replica sets

### PR DESCRIPTION
**Pull request by antiguru**
### Motivation

Implement replica sets, as outlined in #21166. Merging pending approval of the design doc.

This change implements:
* [x] `CREATE REPLICA SET pfname` supporting managed and unmanaged replica sets
* [x] `ALTER REPLICA SET pfname` for managed and unmanaged replica sets, owner, name
* [x] `DROP REPLICA SET pfname` for all replica sets, cleaning up dependent resources.
* [x] `SHOW REPLICA SETS`
* [x] `CREATE CLUSTER REPLICA clname.rname IN REPLICA SET pfname` for unamanged replica sets
* [x] `DROP CLUSTER REPLICA clname.rname` for replica set replicas
* [x] `mz_replica_sets` catalog table
* [x] `mz_show_replica_sets` catalog view
* [x] `mz_show_cluster_replicas` with `replica_set_id` column.
* [x] Stash migration
* [x] Does this make `PROFILE` a reserved cluster name? --> No, since the new `CREATE REPLICA SET` syntax doesn't collide with any other syntax.
* [x] Feature flag
* [ ] SLT tests, parser tests
* [ ] Testdrive
* [ ] How should we name replicas of managed replica sets? `{replica_set_name}_r{index}`? This influences renaming replica sets because now all replicas would need to be renamed, too.

### Tips for reviewer

The change is substantial, and I'm not happy about that. It's currently structured in two logical parts:
1. Add support for cluster items. This doesn't change anything user-visible and only prepares the catalog to allow different types of items within cluster objects.
2. Add replica sets as a second variant of cluster items. Includes parsing/planning/sequencing, which is why it's so substantial.

I think the changes are largely mechanical. Here are some questions:
* The parser shares option parsing for clusters and replica sets because they're identical at the moment. This is a smell, but I decided to go this route to not require adding more code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
Since this PR touches quite some code and parser functionality, I would like to run a coverage run to detect possible test gaps. Please inform me once this PR is ready for that. Thank you!

**Comment by nrainer-materialize** [comment]
I have triggered a [coverage build](https://buildkite.com/materialize/coverage/builds/194) and will report the findings once the build has completed.
I will also have a look at the PR.

**Comment by mjibson** [comment]
I'm not convinced we want the first commit. While users might need to have the experience of "clusters and replicas in the implicit replica set cannot share names" (or is it *any* replica set?), that is a thing we should perhaps enforce during planning. That allows us to keep the internal representation simpler than this: cluster has a collection of replica sets, each set has a collection of replicas.

**Comment by nrainer-materialize** [comment]
Thanks for the upgrade and SLT logic tests.

> I have triggered a [coverage build](https://buildkite.com/materialize/coverage/builds/195) (retried due to a problem) and will report the findings once the build has completed. I will also have a look at the PR.

Alright, we got some results after some retries in [#196.](https://buildkite.com/materialize/coverage/builds/196) Please note that not all builds succeeded and that coverage of failed builds is not included (unfortunately, it is not yet possible to update the coverage after retrying individual builds). Therefore, feel free to ignore a finding if you think it does not make sense.

* All lines are covered by at least one test.
* A few lines are covered by unit tests but not end-to-end tests. I will mark them in the code.

Besides that, I tested the feature manually and it works as expected. 👍

In addition, I will add a platform check, which additionally checks if things still work after restarts and other events.

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -1606,11 +1632,69 @@ impl<T: AstInfo> AstDisplay for AlterClusterStatement<T> {
...
+                f.write_str(")");
+            }
+            AlterReplicaSetAction::ResetOptions(options) => {
+                f.write_str("RESET (");
+                f.write_node(&display::comma_separated(options));
+                f.write_str(")");
+            }
+        }
+    }
+}
```
Comment (nrainer-materialize) [inline src/sql-parser/src/ast/defs/statement.rs:1667]: According to coverage data, this method is not covered by any end-to-end test. Would it make sense to add a test for this or is this more about debugging output?

File: src/sql-parser/src/ast/defs/statement.rs [hunk 2]
```diff
@@ -1606,11 +1632,69 @@ impl<T: AstInfo> AstDisplay for AlterClusterStatement<T> {
...
+impl<T: AstInfo> AstDisplay for CreateReplicaSetStatement<T> {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        f.write_str("CREATE REPLICA SET ");
+        f.write_node(&self.of_cluster);
+        f.write_str(".");
+        f.write_node(&self.definition.name);
+        f.write_str(" ");
+        f.write_node(&display::comma_separated(&self.definition.options));
+    }
+}
```
Comment (nrainer-materialize) [inline src/sql-parser/src/ast/defs/statement.rs:1688]: Not covered by end-to-end tests. Same here: Would it make sense to add a test for this or is this more about debugging output?

File: src/sql-parser/src/parser.rs [hunk 3]
```diff
@@ -4086,10 +4119,76 @@ impl<'a> Parser<'a> {
...
+                            .parse_identifier()
+                            .map_parser_err(StatementKind::AlterOwner)?;
+                        let name = UnresolvedObjectName::ClusterItem(name);
+                        Ok(Statement::AlterOwner(AlterOwnerStatement {
+                            object_type,
+                            if_exists,
+                            name,
+                            new_owner,
+                        }))
+                    }
```
Comment (nrainer-materialize) [inline src/sql-parser/src/parser.rs:4143]: Covered by unit tests but not end-to-end tests according to coverage data. Can we add an end-to-end test that covers this code?

File: src/sql-parser/src/parser.rs [hunk 4]
```diff
@@ -4086,10 +4119,76 @@ impl<'a> Parser<'a> {
...
+                            .parse_comma_separated(Parser::parse_cluster_option_name)
+                            .map_parser_err(StatementKind::AlterReplicaSet)?;
+                        self.expect_token(&Token::RParen)
+                            .map_parser_err(StatementKind::AlterReplicaSet)?;
+                        Ok(Statement::AlterReplicaSet(AlterReplicaSetStatement {
+                            if_exists,
+                            name,
+                            action: AlterReplicaSetAction::ResetOptions(names),
+                        }))
+                    }
```
Comment (nrainer-materialize) [inline src/sql-parser/src/parser.rs:4170]: Covered by unit tests but not end-to-end tests according to coverage data. Can we add an end-to-end test that covers this code?

File: src/sql-parser/src/ast/defs/statement.rs [hunk 5]
```diff
@@ -1622,6 +1706,11 @@ impl<T: AstInfo> AstDisplay for CreateClusterReplicaStatement<T> {
         f.write_str(".");
         f.write_node(&self.definition.name);
         f.write_str(" ");
+        if let Some(profile) = &self.of_replica_set {
+            f.write_str("IN REPLICA SET ");
+            f.write_node(profile);
+            f.write_str(" ");
+        }
```
Comment (nrainer-materialize) [inline src/sql-parser/src/ast/defs/statement.rs:1713]: Same here, also not covered by end-to-end tests but not sure if relevant.

**Review by nrainer-materialize — APPROVED** [review]

**Comment by antiguru** [comment]
Closing this as superseded by #21629. If we want to implement replica sets, this can serve as a guide on what needs to change, but it's probably too much of an effort to resolve all the conflicts of this PR.

**Pull request closed without merging**